### DISCUSSION TITLE
Why can't I use `renderToString` anymore? How can I print now?

### DISCUSSION THREAD

**Discussion by JacobWeisenburger** [vercel/next.js#57631] [Help] [answered]
I was using `renderToString` to make some html and then putting that into an iframe to print it. Is there another way to do this?

I was using:

```tsx
import { renderToString } from 'react-dom/server'

const html = renderToString( <SomeReactNode /> )
const iframe = document.createElement( 'iframe' )
iframe.id = 'print_pdf'
iframe.style.width = '100vw'
iframe.style.height = '100vh'
document.body.appendChild( iframe )

iframe.contentWindow?.document.open()
iframe.contentWindow?.document.write( html )
iframe.contentWindow?.document.close()

iframe.contentWindow?.print()
document.body.removeChild( iframe )
```

![image](https://github.com/vercel/next.js/assets/31667350/0b243aac-12f7-4a3c-8298-f10980122e99)

**Comment by blentz100** [comment]
I don't know the answer but the error says to open an issue in this repo, have you done that ?

**Reply by JacobWeisenburger** [reply]
https://github.com/vercel/next.js/issues/57716

it got closed instantly

**Comment by aziyatali** [comment]
As suggested [here](https://react.dev/reference/react-dom/server/renderToString#alternatives)

- If you use Node.js, use [renderToPipeableStream.](https://react.dev/reference/react-dom/server/renderToPipeableStream)
- If you use Deno or a modern edge runtime with [Web Streams](https://developer.mozilla.org/en-US/docs/Web/API/Streams_API), use [renderToReadableStream.](https://react.dev/reference/react-dom/server/renderToReadableStream)

**Reply by JacobWeisenburger** [reply]
I saw this, but it requires me to change a lot to make it work

**Reply by JacobWeisenburger** [reply]
![image](https://github.com/vercel/next.js/assets/31667350/3d975999-59f9-4140-9485-71cd1bad1623)

**Comment by icyJoseph** [comment]
I think if you want to render into an iframe with React, you can do:

```tsx
"use client";
import { PropsWithChildren, useState } from "react";
import { createPortal } from "react-dom";

const CustomIframe = ({ children, ...props }: PropsWithChildren) => {
  const [contentRef, setContentRef] = useState<HTMLIFrameElement | null>(null);

  const mountNode = contentRef?.contentWindow?.document?.body;

  return (
    <iframe {...props} ref={setContentRef}>
      {mountNode && createPortal(children, mountNode)}
    </iframe>
  );
};

export default CustomIframe;
```

And then:

```tsx
import CustomIframe from "./CustomIframe";

export default function Home() {
  return (
      <CustomIframe>
        <h1>Hello World!</h1>
      </CustomIframe>
  );
}
```

If you don't want to do the Portal trick, you might want to try and dynamically import `react-dom/server`:

```tsx
export default async function Home() {
  const { renderToString } = await import("react-dom/server");

  return (
      <div
        dangerouslySetInnerHTML={{
          __html: `${renderToString(<h1>Hello world 2</h1>)}`,
        }}
      />
  );
}

```

**Reply by JacobWeisenburger** [reply]
This looks promising, I'll have to try it out

**Reply by icyJoseph** [reply]
Did you manage to get it working? Which one did you take?

**Reply by JacobWeisenburger** [reply]
I haven't had time to try it out yet. but I will let you know

**Reply by JacobWeisenburger** [reply]
### Reply to suggestion 1:
I was hoping for a way to print:
1. while only showing the page that's going to be printed in the print preview dialog, not on the web page
2. from a ui button, not `ctrl + p`
3. without having to restructure my current code, which per say 'exits' React to do the printing logic

---

### Reply to suggestion 2:
![image](https://github.com/vercel/next.js/assets/31667350/4cf6e1b5-ba76-40c9-a813-50b5a1bb596e)

Is there another way to take a React component and turn it into an HTML string. I tried using Preact, but ran into some problems with 'pragma', which I have no idea what that means and so I stopped trying to do that.

**Reply by icyJoseph** [reply]
What Next.js version have you got?

```tsx
export default async function Home() {
  const { renderToString } = await import("react-dom/server");

  return (
    <div
      dangerouslySetInnerHTML={{
        __html: `${renderToString(<h1>Hello world 2</h1>)}`,
      }}
    ></div>
  );
}

```

This works fine for me. No warnings/errors at all. Bootstrapped with create-next-app today, so it got next@14.

**Reply by JacobWeisenburger** [reply]
next: 14.0.0

```ts
"dependencies": {
    "@hookform/resolvers": "^2.9.11",
    "@js-temporal/polyfill": "^0.4.3",
    "date-fns": "^2.30.0",
    "encoding": "^0.1.13",
    "firebase": "^9.20.0",
    "freerstore": "^0.4.0",
    "lodash": "^4.17.21",
    "nanoid": "^5.0.1",
    "next": "^14.0.0",
    "pdf-lib": "^1.17.1",
    "radash": "^10.8.1",
    "react": "18.2.0",
    "react-day-picker": "^8.8.0",
    "react-dom": "18.2.0",
    "react-hook-form": "^7.45.4",
    "sass": "^1.58.3",
    "twind": "^0.16.19",
    "yaml": "^2.2.1",
    "zod": "^3.20.2",
    "zustand": "^4.3.6"
},
"devDependencies": {
    "@types/lodash": "^4.14.191",
    "@types/node": "^18.11.18",
    "@types/react": "^18.0.26",
    "@types/react-dom": "^18.0.10",
    "@typescript-eslint/eslint-plugin": "^5.47.1",
    "@typescript-eslint/parser": "^5.47.1",
    "autoprefixer": "^10.4.7",
    "eslint": "^8.30.0",
    "eslint-config-next": "13.1.6",
    "postcss": "^8.4.14",
    "tailwindcss": "^3.2.0",
    "typescript": "^4.9.4"
},
```

**Reply by icyJoseph** [reply]
Oh I guess you are using a client component 🤔 mmm

**Reply by icyJoseph** [reply]
Does it work to do a visually hidden with also aria-hidden combo on the CustomIframe (approach 1) and render a button next to it, and handle on click, so that you print the iframe ref content?

Or with CSS media queries as OP ended up doing.

**Reply by JacobWeisenburger** [reply, marked as answer]
> Oh I guess you are using a client component 🤔 mmm

Yes, because I'm trying to do it with a button onClick.

---

Thanks for all your help. I was able to find [this video](https://www.youtube.com/watch?v=_06QhvUsx_s), and it worked for me.

**Reply by icyJoseph** [reply]
Oh right, facepalm, haha media queries to the rescue! Awesome, well, not sure what was the answer here, but please don't forget to mark as answered :) if there was one 🤔
